### PR TITLE
exercise 07

### PR DESCRIPTION
**Pull request by Semjasa83**

### CODE REVIEW

File: MarceldeZanna/tests_exercise_07/taskScheduler.test.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  describe('cancelAll()', () => {
+    // TODO: Test that cancelAll() prevents tasks from executing
+    const taskScheduler = new TaskScheduler();
+    const taskFn = jest.fn();
+
+    taskScheduler.scheduleRecurring('test-loop', taskFn, 3000);
+    taskScheduler.cancelAll();
+    jest.advanceTimersByTime(8000);
+
+    expect(taskFn).not.toHaveBeenCalled();
```
Comment (clemenshelm) [inline MarceldeZanna/tests_exercise_07/taskScheduler.test.js]: blocker: Dieser Block hat keinen `test(...)`-Wrapper — der Code steht direkt im `describe`-Callback. Jest führt den `describe`-Body beim *Sammeln* der Tests aus, registriert aber keinen Test. Das `expect(taskFn).not.toHaveBeenCalled()` läuft also außerhalb eines Tests und wird entweder still ignoriert oder als Fehler gemeldet. Fazit: dein cancelAll-Test existiert faktisch nicht, obwohl der Code da steht.

Lösung: alles in ein `test(...)` wrappen:

```suggestion
    test('cancelAll prevents scheduled tasks from executing', () => {
      const taskScheduler = new TaskScheduler();
      const taskFn = jest.fn();

      taskScheduler.scheduleRecurring('test-loop', taskFn, 3000);
      taskScheduler.cancelAll();
      jest.advanceTimersByTime(8000);

      expect(taskFn).not.toHaveBeenCalled();
    });
```

Kleiner Tipp, wie du so etwas sofort erkennst: wenn du `npm test` laufen lässt und der cancelAll-Block gar nicht in der Test-Liste auftaucht (kein Häkchen), fehlt der `test`-Wrapper.

**Review by clemenshelm — CHANGES_REQUESTED** [review]
@Semjasa83, gutes Fundament — fake timers in `beforeEach`/`afterEach` korrekt,
`schedule()` mit der „nicht vor dem Delay"-Prüfung über zwei `advanceTimers`
-Schritte sauber aufgebaut, und `scheduleRecurring()` zeigt das Wiederholen
mit dem 1500/1000/500-Raster schön. AAA-Struktur hast du konsequent drin.

blocker: Der `cancelAll()`-Block (Zeile 73–80) hat **keinen `test()`-Wrapper**
— siehe Inline-Kommentar mit der Korrektur. Der Code steht direkt im
`describe`-Body und läuft so nicht als Test. Das ist der häufigste Stolperer
bei describe/test: Code im describe wird nur gesammelt, nicht ausgeführt.

blocker: Zwei Edge Cases aus der Spec fehlen noch:

- **Zero delay (`0ms`)** — was passiert bei `schedule('x', fn, 0)`? Erwartet:
  Task läuft sofort (beim nächsten Tick). Test: schedule mit 0ms, advance 0
  oder微量, prüfen dass `fn` aufgerufen wurde.
- **Multiple tasks gleichzeitig** — schedule zwei Tasks mit verschiedenen
  Delays, advance über beide hinweg, prüfen dass beide in `getExecutedTasks()`
  landen (in der richtigen Reihenfolge).

Die sind beide schnell geschrieben — je ein Test im `schedule()`-Block.

praise: Der „nicht vor dem Delay"-Test (Zeile 33–34, erst 4000ms advance,
dann erst nach weiteren 1000ms auslösen) ist genau die Granularität die
zählt — viele testen nur „nach Delay aufgerufen" und übersehen, dass auch
„vor Delay NICHT aufgerufen" ein eigenes Assert will.

Fix die drei Blocker (cancelAll-Wrapper + zwei Edge-Case-Tests), dann
approve ich.

File: MarceldeZanna/tests_exercise_07/taskScheduler.test.js (hunk 1 above)
Comment (Semjasa83) [reply MarceldeZanna/tests_exercise_07/taskScheduler.test.js]: ups das war mir bei der ersten schon passiert

**Review by clemenshelm — APPROVED** [review]
@Semjasa83, alles drin — die drei Blocker sind sauber gefixt:

- `cancelAll()` jetzt korrekt in `test(...)` gewrappt.
- Edge Case Zero delay: `schedule(..., 0)` + `advanceTimersByTime(0)` →
  Task feuert sofort.
- Edge Case Multiple tasks: drei Tasks mit gleichem Delay, nach dem
  Advance alle drei aufgerufen — schön, dass du für jeden einen eigenen
  `jest.fn()` nimmst statt einen zu teilen, sonst könntest du nicht
  unterscheiden wer wie oft dran war.

praise: Dein „Persönlicher Vermerk zur eigenen Dummheit"-Kommentar bei den
multiplen Tasks hat mich zum Schmunzeln gebracht — genau diese
C&P-mit-gleichem-Namen-Falle kennt jeder. Solche ehrlichen Notizen sind
wertvoller als sauberer Code, der vorgibt, von Anfang an perfekt gewesen zu
sein. (Für einen produktiven PR würd' ich den am Ende rausnehmen, hier in
der Übung darf er bleiben.)

LG, das merge ich.
